### DISCUSSION TITLE
[Subtype] Fail of the registration service when service is a subtype

### DISCUSSION THREAD

**Discussion by estelleOl** [openthread/openthread#8248] [Q&A] [answered]
Hello,

I am trying to create a subtype service, I write the serviceName with the subtype name before in ordre to create the subype but the registration of my service fail when I put the sub type label in the nameService. The srpClientCallback return 6. What am I doing wrong ?
Here my code to create the subTybe service : 

       /* Get buffer instance name */
       serviceInst = otSrpClientBuffersGetServiceEntryInstanceNameString(entry, &size);
     
        strcpy(serviceInst, "board2Instance");
        
        /* Get buffer sub type labels array */
        subTypeArray = otSrpClientBuffersGetSubTypeLabelsArray(entry, &size);
        

        /* Set subType Label */
        subTypeArray[0] = "subtypeBoard2\0";
        
        /* Get service name buffer */
        serviceName = otSrpClientBuffersGetServiceEntryServiceNameString(entry, &size);

        strcpy(serviceName, "_subtypeBoard2._sub._testBoard2._udp");
        
        /* Get buffer txt entry */
        txtEntry = otSrpClientBuffersGetServiceEntryTxtBuffer(entry, &size);

        /* Set txtEntry */
        *txtEntry = 123;
        
        /* Set weight port priority service*/
        entry->mService.mPort = 111;
        entry->mService.mPriority = 1;
        entry->mService.mWeight = 1;
        
        /* Add a service to be registered with server */
        status = otSrpClientAddService(NULL, &entry->mService);

**Comment by jwhui** [comment, marked as answer]
The service name and sub types should be configured separately. It appears you are including the sub type in both.

Also, make sure to set the last entry in the sub type array to NULL to indicate the end of the sub type list.

The SRP Client CLI provides an example of configuring sub types:

https://github.com/openthread/openthread/blob/2b726ef29f5c6a58a30973d7b8defaa94a4dc4c4/src/cli/cli_srp_client.cpp#L381-L406

**Reply by estelleOl** [reply]
I modify my code to create a subtype separately : 
 /* Get buffer instance name */
   serviceInst = otSrpClientBuffersGetServiceEntryInstanceNameString(entry, &size);
 
    strcpy(serviceInst, "board2Instance");
    
    /* Get buffer sub type labels array */
    subTypeArray = otSrpClientBuffersGetSubTypeLabelsArray(entry, &size);
    

    /* Set subType Label */
    subTypeArray[0] = "subtypeBoard2";
    subTypeArray[1] = NULL;
    
    /* Get service name buffer */
    serviceName = otSrpClientBuffersGetServiceEntryServiceNameString(entry, &size);

    strcpy(serviceName, "_testBoard2._udp");
    
    /* Set weight port priority service*/
    entry->mService.mPort = 111;
    entry->mService.mPriority = 1;
    entry->mService.mWeight = 1;
    
    /* Add a service to be registered with server */
    status = otSrpClientAddService(NULL, &entry->mService);

The service ion srp client side is well registered and srpClientCallback return no errors and t he good info of my service.
But when the srp server get this service and ask if this service is a sub type service with this methods : OtSrpServerServiceIsSubType the method return false not a subtype.
I have look in the source code it seems that this method return true if in the name of the srp server service there is the string "._sub." in my case the service name is "_testBoard2._udp" without "._sub." but I do not know how to create a srp server subtype service.
I was expecting when I declare subtypesLabel in the service to have a subtype service.
What have I misunderstood ?

**Reply by abtink** [reply]
Sub-type service always accompany base service.  On server side, they show up as multiple services `otSrpServerService` associated with same host (one for base and one for each sub-type). My guess is you are probably only checking the base service. Check `otSrpServerHostFindNextService()` and its related `OT_SRP_SERVER_FLAGS`.

Some suggestions:
- I would recommend reading the all OT APIs documentation (e.g. in `inlcude/openthread/srp_server.h`) to get familiar with how to use the API (or can check it on `openthread.io` site).
- You can look at OT CLI source code as an example of how the OT APIs are used (e.g. check the implementation of `template <> otError SrpServer::Process<Cmd("service")>(Arg aArgs[])` which prints all registered services on server).

**Reply by estelleOl** [reply]
Thank you for your answer it resolve my issue.
